src/cprover/chc_db.h

```diff
@@ -0,0 +1,233 @@
+    if (can_cast_expr<implies_exprt>(m_chc.where()))
+    {
+      return &to_implies_expr(m_chc.where()).op0();
+    }
+    return &m_chc.where();
+  }
+
+  const exprt* head() const {
+    if (can_cast_expr<implies_exprt>(m_chc.where()))
+    {
+      return &to_implies_expr(m_chc.where()).op1();
+    }
+    return nullptr;
+  }
+
+  bool is_fact() const {
+    auto b = body();
+    bool not_fact = false;
+    b->visit_pre(
+      [&not_fact](const exprt &expr)  {
+        if(can_cast_expr<function_application_exprt>(expr))
+        {
+          not_fact = true;
+        }
+      });
+    return !not_fact;
+  }
+
+  bool is_query() const {
+    if (can_cast_expr<implies_exprt>(m_chc.where()))
+    {
+      auto h = head();
+      bool res = true;
+      h->visit_pre(
+        [&res](const exprt &expr) {
+          if(can_cast_expr<function_application_exprt>(expr))
+            res = false;
+      });
+      return res;
+    }
+    return false;
+  }
+
+  bool operator==(const horn_clauset &other) const
+  {
+    return m_chc == other.m_chc;
+  }
+
+  bool operator!=(const horn_clauset &other) const
+  {
+    return !(*this==other);
+  }
+
+  bool operator<(const horn_clauset &other) const
+  {
+    return m_chc < other.m_chc;
+  }
+
+  void used_relations(chc_dbt &db, std::vector<symbol_exprt> & out) const;
+  void used_func_app(chc_dbt &db, std::vector<function_application_exprt> & out) const;
+};
+
+/*
+ * A database of CHCs.
+ * Uninterpreted relations need to be registered.
+ */
+class chc_dbt
+{
+  friend class horn_clauset;
+public:
+  struct is_state_pred : public std::__unary_function<exprt, bool> {
+    const chc_dbt &m_db;
+    is_state_pred(const chc_dbt &db) : m_db(db) {}
+
+    bool operator()(symbol_exprt state) { return m_db.has_state_pred(state); }
+  };
+
+  typedef std::unordered_set<std::size_t> chc_sett;
+
+private:
+  using chcst = std::vector<horn_clauset>;
+  chcst m_clauses;
+
+  std::unordered_set<symbol_exprt, irep_hash> m_state_preds;
+
+  typedef std::map<exprt, chc_sett> chc_indext;
+  chc_indext m_body_idx;
+  chc_indext m_head_idx;
+
+  // representing the empty set
+  static chc_sett m_empty_set;
+
+public:
+  chc_dbt() {}
+
+  void add_state_pred(const symbol_exprt & state) { m_state_preds.insert(state); }
+  const std::unordered_set<symbol_exprt, irep_hash> &get_state_preds() { return m_state_preds; }
+  bool has_state_pred(const symbol_exprt & state) const { return m_state_preds.count(state) > 0; }
+
+  void build_indices();
+  void reset_indices();
+
+  const chc_sett & use(const exprt & state) const {
+    auto it = m_body_idx.find(state);
+    if (it == m_body_idx.end())
+      return m_empty_set;
+    return it->second;
+  }
+
+  const chc_sett & def(const exprt & state) const {
+    auto it = m_head_idx.find(state);
+    if (it == m_head_idx.end())
+      return m_empty_set;
+    return it->second;
+  }
+
+  void add_clause(const forall_exprt & f)
+  {
+    if (f.is_true())
+      return;
```

Comment:
Can this ever be the case? `f` is a `forall_exprt` (and not a forall-expression-processed-by-the-simplifier).